protocols/streaming/src/lib.rs

```diff
@@ -0,0 +1,547 @@
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        tracing::trace!("new_handler");
+        StreamingProtocolsHandler::new(self.next_inbound_id.clone(), self.config.clone())
+    }
+
+    fn addresses_of_peer(&mut self, peer: &PeerId) -> Vec<Multiaddr> {
+        let mut addresses = Vec::new();
+        if let Some(connections) = self.connected.get(peer) {
+            addresses.extend(connections.iter().filter_map(|c| c.address.clone()))
+        }
+        if let Some(more) = self.addresses.get(peer) {
+            addresses.extend(more.into_iter().cloned());
+        }
+        addresses
+    }
+
+    fn inject_connected(&mut self, peer: &PeerId) {
+        if let Some(pending) = self.pending_outbound_requests.remove(peer) {
+            for request in pending {
+                let sent_request = self.try_open(*peer, request);
+                assert!(sent_request);
+            }
+        }
+    }
+
+    fn inject_connection_established(
+        &mut self,
+        peer: &PeerId,
+        conn: &ConnectionId,
+        endpoint: &ConnectedPoint,
+    ) {
+        let address = match endpoint {
+            ConnectedPoint::Dialer { address } => Some(address.clone()),
+            ConnectedPoint::Listener { .. } => None,
+        };
+        self.connected
+            .entry(*peer)
+            .or_default()
+            .push(Connection::new(*conn, address));
+    }
+
+    fn inject_connection_closed(
+        &mut self,
+        peer_id: &PeerId,
+        conn: &ConnectionId,
+        _: &ConnectedPoint,
+    ) {
+        let connections = self
+            .connected
+            .get_mut(peer_id)
+            .expect("Expected some established connection to peer before closing.");
+
+        let connection = connections
+            .iter()
+            .position(|c| &c.id == conn)
+            .map(|p: usize| connections.remove(p))
+            .expect("Expected connection to be established before closing.");
+
+        if connections.is_empty() {
+            self.connected.remove(peer_id);
+        }
+
+        for id in connection.pending_inbound {
+            self.pending_events
+                .push_back(NetworkBehaviourAction::GenerateEvent(
+                    StreamingEvent::InboundFailure {
+                        peer_id: *peer_id,
+                        id,
+                        error: InboundFailure::ConnectionClosed,
+                    },
+                ));
+        }
+
+        for id in connection.pending_outbound {
+            self.pending_events
+                .push_back(NetworkBehaviourAction::GenerateEvent(
+                    StreamingEvent::OutboundFailure {
+                        peer_id: *peer_id,
+                        id,
+                        error: OutboundFailure::ConnectionClosed,
+                    },
+                ));
+        }
+    }
+
+    fn inject_disconnected(&mut self, peer: &PeerId) {
+        self.connected.remove(peer);
```

Comment:
Given that `inject_connection_closed` above removes a peer from `connected` if no more connections exist and given that `inject_disconnected` is only ever called after `inject_connection_closed, this line should never be executed, correct? If so, how about at least adding a `debug_assert` here to make sure `inject_connection_closed` will properly clean up in the future.

https://github.com/libp2p/rust-libp2p/blob/98bc5e64869d0d188384a91ba9287eff4eb27b0e/swarm/src/lib.rs#L579-L583